x-pack/plugins/lens/public/editor_frame_service/editor_frame/data_panel_wrapper.tsx

```diff
@@ -45,15 +46,18 @@ export const DataPanelWrapper = memo((props: DataPanelWrapperProps) => {
     : true;
 
   const dispatchLens = useLensDispatch();
-  const setDatasourceState: StateSetter<unknown> = useMemo(() => {
-    return (updater) => {
+  const setDatasourceState: StateSetter<unknown, { applyImmediately?: boolean }> = useMemo(() => {
+    return (updater: unknown | ((prevState: unknown) => unknown), options) => {
       dispatchLens(
         updateDatasourceState({
           updater,
           datasourceId: activeDatasourceId!,
           clearStagedPreview: true,
         })
       );
+      if (options?.applyImmediately) {
+        dispatchLens(applyChanges());
```

Comment:
I am wondering if it wouldn't be more performant if instead of dispatching 2 actions one after another, we could add a parameter `options` to `updateDatasourceState`? So we'd invoke it  in this way
` ``
dispatchLens(
        updateDatasourceState({
          updater,
          datasourceId: activeDatasourceId!,
          options: {
               ...options,
               clearStagedPreview: true,
        })
      );`

and then we'd incorporate the logic of `applyChanges` action inside `updateDatasourceState`. Same for other places when we run applyChanges right after running another actions (I think there are 3 in this PR) I didn't test it though so not sure if that would work ok. WDYT?

Reply:
Thanks for the feedback! I went the direction of two dispatches to keep the reducers simple and atomic. Can you help me understand in what way using a single `dispatch` call would be more performant?

Reply:
So main my concern is that we dispatch another action so we have another rerender of the app. I might be wrong though, I might test it on Monday :D Here's a random article that explains it: https://unsplash.com/blog/react-redux-performance-considerations-when-dispatching-multiple-actions/

Reply:
Did you get a chance to test this? Not your responsibility, just wanted to check if it has been done.